packages/api/src/client/util.ts

```diff
@@ -1,13 +1,35 @@
 /**
  * GENERATED CODE - DO NOT MODIFY
  */
-export function isObj(v: unknown): v is Record<string, unknown> {
-  return typeof v === 'object' && v !== null
+export type $Type<Id extends string, Hash extends string> = Hash extends 'main'
+  ? Id | `${Id}#${Hash}`
+  : `${Id}#${Hash}`
+
+function has$type<V>(v: V): v is V & object & { $type: unknown } {
+  return v != null && typeof v === 'object' && '$type' in v
+}
+
+function check$type<Id extends string, Hash extends string>(
+  $type: unknown,
+  id: Id,
+  hash: Hash,
+): $type is $Type<Id, Hash> {
+  return (
+    typeof $type === 'string' &&
+    ($type === id
+      ? hash === 'main'
+      : // $type === `${id}#${hash}`
+        $type.length === id.length + 1 + hash.length &&
+        $type[id.length] === '#' &&
+        $type.startsWith(id) &&
+        $type.endsWith(hash))
```

Comment:
For workloads where `check$type()` is usually `true` it's not clear to me that this is going to have the intended effect of a performance improvement over the simpler version.  Or is the idea specifically to avoid allocating the additional string?

Reply:
The primary goal is to avoid string allocation.

The fact that the order of operation can be optimized for the `false` case is an added benefit.